### PR TITLE
Migrate to UAVCAN v1

### PR DESCRIPTION
**Pull request by davidlenfesty**
Given that libcanard (what we use on the embedded side) and pyuavcan both have release support for UAVCAN v1, I would like to migrate.

This is the intro work we need to migrate to UAVCAN v1, but can't be tested until ROS supports Python 3.7. We *could* build a docker container that builds ROS melodic for Python 3.7, or we could wait until Noetic comes out and migrate wholesale because screw Python 2.

It doesn't make since for the embedded side to make the migration to V1 until software supports it, so this is blocking for the embedded migration.

### CODE REVIEW

**Comment by davidlenfesty** [comment]
I'm really close to getting the build system to where I want it. I'm just having two issues with it. (I'll work on it more later too but maybe this is easier to fix for someone more familiar with the whole setup).

1. If I leave the final `add_custom_command` in the CMakeLists.txt it fails to build when I'm building the docker image. It complains about `-std=c++14` being invalid for C files. I could probably easily fix this, but I don't get the same error if I remove that section for the docker image build and add it back afterwards.
2. It doesn't want to copy over the submodules for `uavcan_dsdl` properly into the docker image when building. If you build the image with the last `add_custom_command` CMake command commented out, and then add it back in, you get an error that the directory inside of `uavcan_dsdl` doesn't exist.

**Comment by davidlenfesty** [comment]
I circumvented these issues by just putting everything in the scripts. Still need to do some quick testing to make sure everything works how I want it, but then it should be good to go.

**Comment by davidlenfesty** [comment]
Works AFAIK, could do with some cleanup but I'm happy enough with it.

Still left in some extra testing nodes, will remove before merging or anything, after I've tested it again (I tested before the cleanup).

**Comment by mdepp** [comment]
There's some redundancy here with how the drive code is already implemented. How ROS architects this is
  - We send drive commands (twist messages) to the drive controller. This controller is configured [here](https://github.com/UofA-SPEAR/software/blob/david/uavcan-v1/spear_rover/config/controllers.yaml). It essentially converts twist messages on the topic `/drive_controller/cmd_vel` to velocity commands for the wheel links.
  - The [hardware](https://github.com/UofA-SPEAR/software/blob/david/uavcan-v1/spear_rover/src/hardware.cpp) node reads the velocities and sends them over to the firmware... at least that's what it SHOULD be doing. What it ACTUALLY does is send the wheel velocities over to canros.

Since your code is a replacement for canros, I believe the idiomatic ROS way to do what this PR does would be to do a `TX_TRANSFER` in `hardware.cpp` (specifically in `CaseDriveHardware::write` instead of using `wheel_commands_publisher`).  That'll remove a lot of code that's essentially duplicating what the ROS controllers and hardware interface is already doing.

@davidlenfesty what do you think?

**Comment by mdepp** [comment]
Slight correction, I looked over the code some more and it probably won't result in all that much code being removed... essentially all that will happen is the stuff that `mapper.cpp` does will move to `hardware.cpp`, and as a bonus we remove using ROS messages in the _twist message -> wheel velocity commands -> CAN messages_ chain.

**Comment by davidlenfesty** [comment]
I put in drive commands as more of an example than anything. The intent here is to publish/recieve arbitrary CAN messages.

The intent here is for any ROS node to be able to interact with a UAVCAN message as if it was a ROS topic. This includes drive, arm, battery monitoring, any science equipment we add to the rover or whatever. If every node that had to deal with CAN was directly interfacing with the bus, then all those nodes would have to implement the UAVCAN node. I suppose that's very feasible, especially since UavcanMapper is fairly general.

I was trying to reduce the ROS <-> CAN interface to a single node, and keep all that translation in a single spot.

### Per-node UAVCAN

#### Pros
- Reduces coupling between different nodes, fewer message types to maintain.
- More idiomatic maybe? I don't know what idiomatic ROS is honestly
- System isn't dependent on a single node being started correctly and responding to everything correctly.

#### Cons
- Every node has to handle every CAN frame to make sure it catches what it cares about. (not a huge deal, CAN is only running at 250kbps and nowhere near saturated).
- Every node has to deal with also initializing the mapping interface, *and* has to be configured for the correct CAN interface

### Single mapper node
#### Pros
- Single place to manage UAVCAN/ROS bridge
- ROS nodes can just deal with ROS, no extra semantics to be aware of
- One thread to parse incoming CAN messages and dispatch to ROS system.
#### Cons
- dependent on a single node being started and operating correctly.
- Seperate place to manage some messages, if you make a change to some ROS->CAN message in a node you have to propogate those to another node (the mapper)

TL;DR is moving stuff into hardware.cpp either means that becomes the new mapper node or that any node that wants access to CAN bus needs to manage it directly. The latter could have benefits but I would still prefer we keep a single mapper node.

**Comment by mdepp** [comment]
I don't like the idea of the hardware interface node having to interact with the hardware through ROS messages, but I also don't really like the idea of hardware.cpp just becoming another mapper (for battery messages etc.). Maybe an ideal solution would be to have two CAN communication nodes, one for hardware.cpp and one for miscellaneous stuff?

How about for now we just update `hardware.cpp` to point to the new mapper instead of to canros. Then we can improve it later if necessary.

File: spear_rover/src/mapper.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+// TODO error handling
+#define RX_TRANSFER(type, xfer, out_msg) {\
+  size_t _deser_size = xfer->payload_size; \
+  type ## _deserialize_(&out_msg, static_cast<const uint8_t*>(xfer->payload), &_deser_size); \
+}
+
+/* -------- Main mapper helpers -------- */
+// TODO improve visibility constraints
+struct Can2Ros {
```
Comment (mdepp) [inline spear_rover/src/mapper.h:47]: I can't figure out what this thing actually is... like if I look at how it's used in `mapper.cpp` it's this:

```
Can2Ros sub{};
sub.subject_id = 0;
sub.callback = can2ros_odom_cb;
m->map_can2ros(sub);
sub.subject_id = 1;
sub.callback = can2ros_arm_cb;
m->map_can2ros(sub);
```
My best guess is that this is a combination of function arguments, and storage for the canard subscription. In particular I notice that the *same* subscription is used each time. Is there ever a case where this shouldn't happen? (b/c if not `subscription` should be a member variable of `UavcanMapper`).

File: spear_rover/src/mapper.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  CanardPortID subject_id;
```
Comment (mdepp) [inline spear_rover/src/mapper.h:48]: looks like it should be called `port_id`

File: spear_rover/src/mapper.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  // Transfer ID must monotonically increase
+  static uint8_t transfer_id = 0;
+
+  // Business conversion logic, in this case it's a 1:1 mapping so very simple.
+  spear_actuators_DriveCommand_1_0 cmd;
+  cmd.speed = msg->speed;
+  cmd.id.value = msg->id;
+
+  // Simplest I could make a macro.
+  TX_TRANSFER(spear_actuators_DriveCommand_1_0, cmd, transfer_id, 0, CanardPriorityNominal);
```
Comment (mdepp) [inline spear_rover/src/mapper.cpp:30]: ...

You almost slipped this one past me, but this macro implicitly uses the global variable `m` doesn't it? This definitely needs to change. You could pass `m` into `TX_TRANSFER`, but I think a better idea is to move some of this logic into `m` itself if possible.

File: spear_rover/src/can_sender.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
+#include <ros/ros.h>
```
Comment (mdepp) [inline spear_rover/src/can_sender.cpp:1]: Is this just demo/test code? I'd rather keep that to a minimum, but if it's useful can you add a quick comment at the top here explaining its purpose?

File: spear_rover/src/hardware.cpp [hunk 5]
```diff
@@ -64,33 +60,30 @@ class CASEDriveHardware : public RobotHWChild {
...
-        "/canros/msg/uavcan/equipment/actuator/ArrayCommand", 1);
-    boost::function<void(const WheelOdomPtr &)> odom_callback =
-        [&](const WheelOdomPtr &message) { on_wheel_odom_message(message); };
+    wheel_commands_publisher = nh.advertise<spear_msgs::DriveCommand>(
+        "/can/spear/actuators/drive_command", 1);
+    boost::function<void(const spear_msgs::DriveOdometry::ConstPtr &)> odom_callback =
+        [&](const spear_msgs::DriveOdometry::ConstPtr &message) { on_wheel_odom_message(message); };
     wheel_odom_subscriber =
-        nh.subscribe("/canros/msg/spear/drive/WheelOdom", 1000, odom_callback);
+        nh.subscribe("/canros/msg/spear/drive/drive_odometry", 1000, odom_callback);
```
Comment (mdepp) [inline spear_rover/src/hardware.cpp:68]: Looks like it should be `"/can/spear/<something>"`

**Review by mdepp — CHANGES_REQUESTED** [review]
Also I have a general question: Does this fully supersede `mapper.py` And if not, is it meant to eventually do so?

File: spear_rover/src/mapper_helpers.cpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+  // TODO anonymous nodes are unsupported for now
+  can_node.node_id = can_id;
+
+  // Note: I was talking with some external people about some of the C++ stuff,
+  // and substrate was recommended to me as a library: https://github.com/bad-alloc-heavy-industries/substrate
+  //
+  // It has much nicer abstractions for sockets and some core stuff that would be much nicer to use.
+  // Integrating it could be as simple as adding it as a submodule and an include directory, but it's an
+  // extra unnecessary dependancy, so I'm not adding it in yet unless others find it interesting.
```
[the remark this answers is not in the document]
Comment (davidlenfesty) [reply spear_rover/src/mapper_helpers.cpp:16]: Sounds good. The terminology I tend to see is the folder `vendor`, and yeah I'll make a new PR for this specifically later.

File: spear_rover/src/mapper.h (hunk 2 above)
Comment (davidlenfesty) [reply spear_rover/src/mapper.h:48]: Yeah generically it's port ID but for messages it's subject ID, and we're only handling messages, not services here, I got confused :). Will change.

File: spear_rover/src/mapper.h (hunk 1 above)
Comment (davidlenfesty) [reply spear_rover/src/mapper.h:47]: Yeah really the way I'm using this it should really just be function arguments. Unless I'm misinterpreting some C++ semantics (this is how it would work in C), I'm copying the Can2Ros object, which is then pushed into a vector of them inside of mapper, so making a struct and passing it into map_can2ros is just wasting lines of code.

Can2Ros shouldn't be visible to the user and this usage is just vestigial code from how I thought it was going to work when I started.

File: spear_rover/src/mapper.h [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  std::function<void(CanardTransfer*)> callback;
+  CanardRxSubscription subscription;
+};
+
+class UavcanMapper {
+  public:
+    UavcanMapper(CanardNodeID can_id, std::string can_iface);
+    void map_can2ros(Can2Ros sub);
+    void handle_frame(struct can_frame* in_frame);
```
[the remark this answers is not in the document]
Comment (davidlenfesty) [reply spear_rover/src/mapper.h:57]: `handle_frame` will get moved for sure, `send_transfer` is addressed in the comment below about `TX_TRANSFER`.

File: spear_rover/src/mapper.h (hunk 1 above)
Comment (mdepp) [reply spear_rover/src/mapper.h:47]: No you're right that it copies (assuming the subscription class has correct copy semantics), so yeah it looks like it should just be function arguments.

File: spear_rover/src/mapper.cpp (hunk 3 above)
Comment (davidlenfesty) [reply spear_rover/src/mapper.cpp:30]: Woops, my C is showing.

Yeah this one isn't great, it was more of a matter of convenience than anything. I feel like the best option is to pass `m` in (and probably just declare these as lambda functions, that might make the code easier to read and would avoid using a global `m` - kinda, it still has the same problems but at least it's constrained to `main`). The point was to just make it so people didn't have to worry about allocating a correctly-sized buffer or managing the serialization functions, which I feel would be hampered if we split the functionality out.

In this case I prefer "easy to use" to "technically better" for now, especially given that this will either be replaced with `mapper.py` once we get access to python 3 or have libcanard swapped out for libuavcan once that's v1-ready (which will be soon :tm: ). My plan was going back to pyuavcan for ease of use but since I have this code already I don't mind using libuavcan as well, we can talk about that when either option is available.

(libuavcan will be a modern C++ implementation of UAVCAN vs. libcanard which is focussed on small-footprint C).

File: spear_rover/src/hardware.cpp (hunk 5 above)
Comment (davidlenfesty) [reply spear_rover/src/hardware.cpp:68]: Yeah I'm not opinionated on this I just copied the naming I was using for the DSDL (which should also maybe just be re-organized). I think maybe there should be a quick discussion sometime on how we want to namespace message types.

File: spear_rover/src/can_sender.cpp (hunk 4 above)
Comment (davidlenfesty) [reply spear_rover/src/can_sender.cpp:1]: Yeah this is just demo code, I'll pull it out.

**Comment by davidlenfesty** [comment]
I responded to some of the comments, will actually make the changes later.

File: spear_rover/src/hardware.cpp (hunk 5 above)
Comment (davidlenfesty) [reply spear_rover/src/hardware.cpp:68]: Oh looking over code comments at 11 may not be the best idea, I see the real problem now.

File: spear_rover/src/mapper.cpp [hunk 8]
```diff
@@ -11,103 +13,82 @@
...
+      cmd.speed = msg->speed;
+      cmd.id.value = msg->id;
+
+      // Simplest I could make a macro.
+      TX_TRANSFER(m, spear_actuators_DriveCommand_1_0, cmd, transfer_id, 0, CanardPriorityNominal);
+      transfer_id++;
+  };
+  auto _ros2can_drive_sub = nh->subscribe("/can/spear/actuators/drive_command", 100, ros2can_drive_cb);
+
+  boost::function<void(const spear_msgs::JointCommand::ConstPtr&)> ros2can_joint_cb =
```
Comment (mdepp) [inline spear_rover/src/mapper.cpp:42]: I think it'd be cleaner to keep use global callback functions like
```
void ros2can_joint_cb(std::shared_ptr<UavcanMapper>& m, const spear_msgs::JointCommand::ConstPtr& msg) { ... }
```

and then use a small wrapper lambda like
```
auto _ros2can_arm_sub = nh->subscribe("/can/spear/actuators/joint_command", 100, [m](const auto msg) {
  ros2can_joint_cb(m, msg);
});
```

(you might have to change this syntax or define the lambda beforehand, I haven't tested it and I'm not sure what C++ version we are working with)

File: spear_rover/src/mapper_helpers.cpp [hunk 9]
```diff
@@ -30,14 +30,17 @@ UavcanMapper::UavcanMapper(CanardNodeID can_id, std::string can_iface) {
 
 }
 
-void UavcanMapper::map_can2ros(Can2Ros sub) {
-  can2ros.push_back(sub);
+void UavcanMapper::map_can2ros(CanardPortID port_id, std::function<void(CanardTransfer*)> callback) {
+  Can2Ros subscription{};
```
Comment (mdepp) [inline spear_rover/src/mapper_helpers.cpp:34]: From this line, I think `Can2Ros` should be renamed to `CanSubscription` (or maybe `Can2RosSubscription`?)

**Review by mdepp — CHANGES_REQUESTED** [review]

**Review by mdepp — APPROVED** [review]
